### PR TITLE
Replace AMR values with the ones sent by the IDPs

### PR DESCRIPTION
**Pull request by mpmadhavig**
$subject

Fixes https://github.com/wso2/product-is/issues/17287

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -3946,10 +3946,64 @@ private void associateAuthenticationHistory(SessionDataCacheEntry resultFromLogi
...
+                OAuthConstants.READ_AMR_VALUE_FROM_IDP));
+        if (readAMRValueFromIdp) {
+            List<String> resultantAuthMethods = new ArrayList<>();
+            Object[] idpKeySet = authenticatedIdPs.keySet().toArray();
+            for (int i = 0; i < authMethods.size(); i++) {
+                boolean amrFieldExists = false;
+                if (idpKeySet[i] != null) {
+                    String idpKey = (String) idpKeySet[i];
+                    for (Map.Entry<ClaimMapping, String> entry : authenticatedIdPs.get(idpKey).getUser()
+                            .getUserAttributes().entrySet()) {
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:3973]: possible NPE

**Comment by nilasini** [comment]
Please give samples before and after your fix in the description

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 1 above)
Comment (mpmadhavig) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Resolved with 79aa8d8d027e01d3774d62ebd20f7f3777cabea0

**Comment by mpmadhavig** [comment]
> Please give sample AMR values before and after your fix in the description

**Resident IDP**
- SP sign in method configured as MFA with first step as the resident IDP's basic authenticator and 2nd step as Federated IDP.

**Federated IDP**
- SP configured with Google Authenticator as the authentication method.

AMR values before and after the fix will be displayed as below.

**Before**
```
"amr": [
    "OpenIDConnectAuthenticator",
    "BasicAuthenticator"
  ]
```

**After**
```
"amr": [
    "GoogleOIDCAuthenticator",
    "BasicAuthenticator"
  ]
```

***To apply the new behaviour add the following config to deployment.toml

```
[replace_amr_value_with_idp_sent_values]
enable="true"
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 2]
```diff
@@ -3946,10 +3946,67 @@ private void associateAuthenticationHistory(SessionDataCacheEntry resultFromLogi
...
+                                break;
+                            }
+                        }
+                    }
+                }
+                if (!amrFieldExists) {
+                    resultantAuthMethods.add(authMethods.get(i));
+                }
+            }
+
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:3988]: remove extra line

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 3]
```diff
@@ -3946,10 +3946,67 @@ private void associateAuthenticationHistory(SessionDataCacheEntry resultFromLogi
...
     }
 
+    /**
+     * Replaces the authenticator names with the AMR values sent by the IDP.
+     *
+     * @param authMethods       The list of authentication methods set by resident IDP.
+     * @param authenticatedIdPs The authenticated IDPs.
+     */
+    private List<String> getAMRValues(List<String> authMethods, Map<String, org.wso2.carbon.identity
+            .application.authentication.framework.model.AuthenticatedIdPData> authenticatedIdPs) {
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:3961]: can't we import and use AuthenticatedIdPData?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 4]
```diff
@@ -3946,10 +3946,67 @@ private void associateAuthenticationHistory(SessionDataCacheEntry resultFromLogi
...
+
+    /**
+     * Adds the authentication methods to the list.
+     *
+     * @param amrValue             Comma separated authentication method value or values.
+     * @param resultantAuthMethods The resultant list of authentication methods.
+     */
+    private void addToAuthMethods(String amrValue, List<String> resultantAuthMethods) {
+
+        if (amrValue.contains(",")) {
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:4002]: check null for amrValue

Comment (mpmadhavig) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:4002]: The value only be set if the amr field is not null. Therefore amr value can never be null.

Comment (mpmadhavig) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:4001]: Fixed with 986056862998a05be12dba4c5697d91175210a9a

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 3 above)
Comment (mpmadhavig) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed with 986056862998a05be12dba4c5697d91175210a9a

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 2 above)
Comment (mpmadhavig) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed with 986056862998a05be12dba4c5697d91175210a9a

**Review by nilasini — DISMISSED** [review]

**Review by nilasini — APPROVED** [review]

**Pull request merged**